### PR TITLE
mcs: fix rule sync when meet "no rule left" and concurrency

### LINKED ISSUE #7418: mcs: inconsistency in watch placement rule
### What did you do?

run SetPlacementRules test in api mode and pd mode

### What did you expect to see?

it is consistency

### What did you see instead?

Before SetPlacementRules
`[2023/11/23 11:59:59.383 +08:00] [INFO] [rule.go:433] ["get placement rules"] [bundles="[{\"group_id\":\"pd\",\"group_index\":0,\"group_override\":false,\"rules\":[{\"group_id\":\"pd\",\"id\":\"default\",\"start_key\":\"\",\"end_key\":\"\",\"role\":\"voter\",\"is_witness\":false,\"count\":3}]}]"]`

After SetPlacementRules in PD mode
`[2023/11/23 11:57:14.503 +08:00] [INFO] [rule.go:433] ["get placement rules"] [bundles="[{\"group_id\":\"5\",\"group_index\":5,\"group_override\":false,\"rules\":[{\"group_id\":\"5\",\"id\":\"foo\",\"index\":1,\"start_key\":\"\",\"end_key\":\"\",\"role\":\"voter\",\"is_witness\":false,\"count\":1,\"create_timestamp\":1700711834}]}]"]`

After SetPlacementRules in API mode
`[2023/11/23 11:54:05.863 +08:00] [INFO] [api.go:959] ["get placement rules"] [bundles="[{\"group_id\":\"pd\",\"group_index\":0,\"group_override\":false,\"rules\":[{\"group_id\":\"pd\",\"id\":\"default\",\"start_key\":\"\",\"end_key\":\"\",\"role\":\"voter\",\"is_witness\":false,\"count\":3}]},{\"group_id\":\"5\",\"group_index\":5,\"group_override\":false,\"rules\":[{\"group_id\":\"5\",\"id\":\"foo\",\"index\":1,\"start_key\":\"\",\"end_key\":\"\",\"role\":\"voter\",\"is_witness\":false,\"count\":1,\"create_timestamp\":1700711616}]}]"]`

### LINKED ISSUE #7452: api: inconsistent results for empty scheduler
### What did you do?

Call API about schedulers when there is no scheduler

### What did you see instead?

before #7131, it returns `[]`

after this, it returns null

### PR DESCRIPTION
**Pull request by lhy1024**
### What problem does this PR solve?

Issue Number: Close #7418

### What is changed and how does it work?

```commit-message
```

1. To address scenarios where deletion precedes an update, we've introduced a pendingDeletion structure. This structure tracks rules or rule groups marked for deletion. If there's an inability to delete immediately due to a lack of rules, it's temporarily stored. Subsequently, during future rule or rule group placement events, the system allows their removal once it becomes feasible due to the availability of required rules.
2. To solve concurrent issues, I merged two watchers that operate on a shared path but handle rules and groups differently. The new watcher now collectively manages the handling of rules and groups, ensuring coordinated and consistent operations in the face of concurrent access. 
3. add more tests.

### Check List

Tests

- Unit test
- Integration test

### Release note

```release-note
None.
```

### CODE REVIEW

File: tests/server/api/rule_test.go [hunk 1]
```diff
@@ -972,20 +975,137 @@ func (suite *ruleTestSuite) checkBundleBadRequest(cluster *tests.TestCluster) {
...
+		func(conf *config.Config, serverName string) {
+			conf.PDServerCfg.KeyType = "raw"
+			conf.Replication.EnablePlacementRules = true
+		},
+	}
+	env := tests.NewSchedulingTestEnvironment(suite.T(), opts...)
+	env.RunTestInTwoModes(suite.checkDeleteAndUpdate)
+}
+
+func (suite *ruleTestSuite) checkDeleteAndUpdate(cluster *tests.TestCluster) {
```
Comment (rleungx) [inline tests/server/api/rule_test.go:989]: Will it fail in the previous implementation?

File: pkg/mcs/scheduling/server/apis/v1/api.go [hunk 2]
```diff
@@ -1330,5 +1330,5 @@ func checkRegionsReplicated(c *gin.Context) {
 		c.String(http.StatusBadRequest, err.Error())
 		return
 	}
-	c.String(http.StatusOK, state)
+	c.IndentedJSON(http.StatusOK, state)
```
Comment (JmPotato) [inline pkg/mcs/scheduling/server/apis/v1/api.go:1333]: What is the reason for this modification? Do we need to modify the corresponding API interface for the non-API mode?

Comment (lhy1024) [reply pkg/mcs/scheduling/server/apis/v1/api.go:1333]: Avoid failure in JSON unmarshalling and maintain consistency with PD mode.

File: pkg/mcs/scheduling/server/rule/watcher.go [hunk 3]
```diff
@@ -62,6 +63,17 @@ type Watcher struct {
 	ruleWatcher  *etcdutil.LoopWatcher
 	groupWatcher *etcdutil.LoopWatcher
```
Comment (rleungx) [inline pkg/mcs/scheduling/server/rule/watcher.go:64]: I still worry about if two watchers will get the event disordered and overwrite the same rule.

Comment (lhy1024) [reply pkg/mcs/scheduling/server/rule/watcher.go:64]: I will add a test about it. If necessary, I will merge these watchers.

File: pkg/mcs/scheduling/server/rule/watcher.go [hunk 4]
```diff
@@ -61,7 +65,6 @@ type Watcher struct {
 	regionLabeler *labeler.RegionLabeler
 
 	ruleWatcher  *etcdutil.LoopWatcher
```
Comment (rleungx) [inline pkg/mcs/scheduling/server/rule/watcher.go:67]: Do we still need it?

Comment (lhy1024) [reply pkg/mcs/scheduling/server/rule/watcher.go:67]: Why not need watcher?

Comment (rleungx) [reply pkg/mcs/scheduling/server/rule/watcher.go:67]: Forget about it, I misread it.

File: pkg/mcs/scheduling/server/rule/watcher.go [hunk 5]
```diff
@@ -103,85 +121,87 @@ func NewWatcher(
...
-		log.Info("delete placement rule", zap.String("key", key))
-		ruleJSON, err := rw.ruleStorage.LoadRule(strings.TrimPrefix(key, prefixToTrim))
-		if err != nil {
-			return err
-		}
-		rule, err := placement.NewRuleFromJSON([]byte(ruleJSON))
-		if err != nil {
-			return err
+		groupID, ruleID, err := func() (string, string, error) {
+			if strings.HasPrefix(string(kv.Key), rw.rulesPathPrefix) {
```
Comment (CabinfeverB) [inline pkg/mcs/scheduling/server/rule/watcher.go:162]:
```suggestion
			if strings.HasPrefix(key, rw.rulesPathPrefix) {
```

File: pkg/mcs/scheduling/server/rule/watcher.go [hunk 6]
```diff
@@ -103,85 +121,87 @@ func NewWatcher(
...
+				if err != nil {
+					return "", "", err
+				}
+				rule, err := placement.NewRuleFromJSON([]byte(ruleJSON))
+				if err != nil {
+					return "", "", err
+				}
+				rw.checkerController.AddSuspectKeyRange(rule.StartKey, rule.EndKey)
+				return rule.GroupID, rule.ID, rw.ruleManager.DeleteRule(rule.GroupID, rule.ID)
+			} else if strings.HasPrefix(string(kv.Key), rw.ruleGroupPathPrefix) {
```
Comment (CabinfeverB) [inline pkg/mcs/scheduling/server/rule/watcher.go:174]: ditto

File: pkg/mcs/scheduling/server/rule/watcher.go [hunk 7]
```diff
@@ -103,85 +121,87 @@ func NewWatcher(
...
+				log.Info("delete placement rule group", zap.String("key", key))
+				trimmedKey := strings.TrimPrefix(key, rw.ruleGroupPathPrefix+"/")
+				for _, rule := range rw.ruleManager.GetRulesByGroup(trimmedKey) {
+					rw.checkerController.AddSuspectKeyRange(rule.StartKey, rule.EndKey)
+				}
+				return trimmedKey, "", rw.ruleManager.DeleteRuleGroup(trimmedKey)
+			} else {
+				log.Warn("unknown key when delete placement rule", zap.String("key", string(kv.Key)))
```
Comment (CabinfeverB) [inline pkg/mcs/scheduling/server/rule/watcher.go:182]: ditto

File: tests/server/api/rule_test.go [hunk 8]
```diff
@@ -925,20 +931,219 @@ func (suite *ruleTestSuite) checkBundleBadRequest(cluster *tests.TestCluster) {
...
+			respBundle := make([]placement.GroupBundle, 0)
+			err = tu.CheckGetJSON(testDialClient, urlPrefix+"/config/placement-rule", nil,
+				tu.StatusOK(re), tu.ExtractJSON(re, &respBundle))
+			suite.NoError(err)
+			if len(respBundle) != len(bundle) {
 				return false
 			}
-		}
-		return true
+			log.Info("respBundle", zap.Any("respBundle", respBundle), zap.Any("bundle", bundle))
```
Comment (rleungx) [inline tests/server/api/rule_test.go:1025]: Will it log too much?

**Review by rleungx — APPROVED** [review]

File: pkg/mcs/scheduling/server/rule/watcher.go [hunk 9]
```diff
@@ -216,3 +238,37 @@ func (rw *Watcher) Close() {
...
+	return len(rw.pendingDeletion.kvs) > 0
+}
+
+func (rw *Watcher) addPendingDeletion(path, groupID, ruleID string) {
+	rw.pendingDeletion.Lock()
+	defer rw.pendingDeletion.Unlock()
+	rw.pendingDeletion.kvs[path] = [2]string{groupID, ruleID}
+}
+
+func (rw *Watcher) tryFinishPendingDeletion() {
```
Comment (rleungx) [inline pkg/mcs/scheduling/server/rule/watcher.go:254]: I am worried about if the put and delete can be disordered. If so, the newly added rule might be deleted unexpectedly.

**Comment by lhy1024** [comment]
We will use replace this pr with txn.

**Pull request closed without merging**